app/assets/javascripts/listings/ListingDataService.js.coffee

```diff
@@ -0,0 +1,434 @@
+    $http.get("/api/v1/listings.json", params).success((data, status, headers, config) ->
+      listings = if data and data.listings then data.listings else []
+      angular.copy(listings, Service.listings)
+      Service.checkFavorites() if checkFavorites
+    ).error( (data, status, headers, config) ->
+      return
+    )
+
+  Service.isAcceptingOnlineApplications = (listing) ->
+    return false if _.isEmpty(listing)
+    return false if ListingLotteryService.lotteryComplete(listing)
+    return false unless ListingIdentityService.isOpen(listing)
+    return listing.Accepting_Online_Applications
+
+  Service.getListingAndCheckIfOpen = (id) ->
+    deferred = $q.defer()
+    Service.getListing(id).then( ->
+      deferred.resolve(Service.listing)
+      if _.isEmpty(Service.listing)
+        # kick them out unless there's a real listing
+        return $state.go('dahlia.welcome')
+      else if !Service.isAcceptingOnlineApplications(Service.listing)
+        # kick them back to the listing
+        return $state.go('dahlia.listing', {id: id})
+    ).catch( (response) ->
+      deferred.reject(response)
+    )
+    deferred.promise
+
+  Service.getListingAMI = ->
+    angular.copy([], Service.AMICharts)
+    Service.loading.ami = true
+    Service.error.ami = false
+    # shouldn't happen, but safe to have a guard clause
+    return $q.when() unless Service.listing.chartTypes
+    allChartTypes = _.sortBy(Service.listing.chartTypes, 'percent')
+    data =
+      'year[]': _.map(allChartTypes, 'year')
+      'chartType[]': _.map(allChartTypes, 'chartType')
+      'percent[]': _.map(allChartTypes, 'percent')
+    $http.get('/api/v1/listings/ami.json', { params: data }).success((data, status, headers, config) ->
+      if data && data.ami
+        angular.copy(Service._consolidatedAMICharts(data.ami), Service.AMICharts)
+      Service.loading.ami = false
+    ).error( (data, status, headers, config) ->
+      Service.loading.ami = false
+      Service.error.ami = true
+      return
+    )
+
+  Service._consolidatedAMICharts = (amiData) ->
+    charts = []
+    amiData.forEach (chart) ->
+      # look for an existing chart at the same percentage level
+      amiPercentChart = _.find charts, (c) -> c.percent == chart.percent
+      if !amiPercentChart
+        # only push chart if it has any values
+        charts.push(chart) if chart.values.length
+      else
+        # if it exists, modify it with the max values
+        i = 0
+        amiPercentChart.values.forEach (incomeLevel) ->
+          chartAmount = if chart.values[i] then chart.values[i].amount else 0
+          incomeLevel.amount = Math.max(incomeLevel.amount, chartAmount)
+          i++
+    charts
+
+  Service.priorityTypes = (listing) ->
+    Service.collectTypes(listing, 'prioritiesDescriptor')
+
+  Service.collectTypes = (listing, specialType) ->
+    _.map listing[specialType], (descriptor) ->
+      descriptor.name
+
+  Service.priorityLabel = (priority, modifier) ->
```

Comment:
can this be moved to ListingConstantsService?

Reply:
It make sense, will do that.